### DISCUSSION TITLE
Difference between mypy and pyright inferring union bounds, which is correct behavior?

### DISCUSSION THREAD

**Discussion by gsingh93** [microsoft/pyright#5040] [Q&A] [answered]
I'm seeing some different behavior between `mypy` and `pyright` with this code on Python 3.11:
```python
from typing import Protocol, Self, TypeVar

class SupportsAdd(Protocol):
    def __add__(self, other: Self, /) -> Self:
        ...

T = TypeVar('T', bound=SupportsAdd)

def add(a: tuple[T, ...], b: tuple[T, ...]) -> tuple[T, ...]:
    return tuple(i + j for i, j in zip(a, b))

s1 = (1, 1)
s2 = (2, 'a')
reveal_type(s1)
reveal_type(s2)
reveal_type(add(s1, s1))
reveal_type(add(s1, s2))
```

The intent of this code was to have an `add` function that could add homogeneous tuples. `pyright` doesn't give an error because it infers a `Union` type where I don't want it to. `mypy` gives me an error like I want, but I'm not sure if it's the "correct" error.

`pyright`:
```
pyright 1.1.305
/Users/gsgx/code/test2.py
  /Users/gsgx/code/test2.py:21:13 - information: Type of "s1" is "tuple[Literal[1], Literal[1]]"
  /Users/gsgx/code/test2.py:22:13 - information: Type of "s2" is "tuple[Literal[2], Literal['a']]"
  /Users/gsgx/code/test2.py:23:13 - information: Type of "add(s1, s1)" is "tuple[int, ...]"
  /Users/gsgx/code/test2.py:24:13 - information: Type of "add(s1, s2)" is "tuple[int | str, ...]"
0 errors, 0 warnings, 4 informations
Completed in 0.67sec
```

`mypy`:
```
test2.py:21: note: Revealed type is "Tuple[builtins.int, builtins.int]"
test2.py:22: note: Revealed type is "Tuple[builtins.int, builtins.str]"
test2.py:23: note: Revealed type is "builtins.tuple[builtins.int, ...]"
test2.py:24: error: Value of type variable "T" of "add" cannot be "object"  [type-var]
test2.py:24: note: Revealed type is "builtins.tuple[builtins.object, ...]"
Found 1 error in 1 file (checked 1 source file)
```

1) Which is the correct behavior in this case?
2) How can I change the signature of the `add` function so that it doesn't infer that `T` is a union? Or maybe the problem here is related to the `SupportsAdd` protocol?

**Comment by erictraut** [comment, marked as answer]
This particular behavior isn't specified in the type specs, so neither type checker is "correct" or "incorrect" here. I think pyright's behavior is arguably more consistent.

Let's consider a simpler example that doesn't involve tuples.

```python
T = TypeVar("T", bound=SupportsAdd)

def add(a: T, b: T) -> T:
    return a + b
```

Let's look at some easy cases where mypy and pyright agree:
```python
reveal_type(add(1, 1)) # Pyright & mypy: int
reveal_type(add("a", "a")) # Pyright & mypy: str
```

Now let's look at a more complex case where mypy and pyright also agree:
```python
def func(x: int | str, y: int | str):
    reveal_type(add(x, y))  # Pyright & mypy: int | str
```

And here's a variant of the complex case where mypy and pyright disagree:
```python
reveal_type(add(1, "a")) # Pyright: int | str, Mypy: error
```

Mypy's constraint solver fails to find a solution that satisfies all of the constraints for `T` in this case. That's because it's using a "join" operator rather than a "union" operator. (This is well-documented [difference between mypy and pyright](https://microsoft.github.io/pyright/#/mypy-comparison?id=constraint-solver-type-widening).) Pyright finds the solution `int | str` that satisfies all of the constraints.

> How can I change the signature of the add function so that it doesn't infer that T is a union?

The job of a constraint solver is to find a solution to the type variables that satisfies all of the constraints. You cannot prevent a constraint solver from widening the type as part of this process. Both "join" and "union" operators are acceptable ways to widen a type.

The only way to prevent widening is to use a constrained TypeVar (that lists all allowed subtypes). That would look something like this:

```python
T = TypeVar("T", int, str)
```

**Reply by gsingh93** [reply]
Thanks!

> Both "join" and "union" operators are acceptable ways to widen a type.

Maybe this is a question for another repo, but I'm wondering if there could be a type hint to ask the constraint solver to choose one or the other for cases like this one.
